internal/internal_nexus_task_handler.go

```diff
@@ -416,3 +422,89 @@ func (p *payloadSerializer) Serialize(v any) (*nexus.Content, error) {
 }
 
 var emptyReaderNopCloser = io.NopCloser(bytes.NewReader([]byte{}))
+
+// convertServiceError converts a serviceerror into a Nexus HandlerError if possible.
+// If exposeDetails is true, the error message from the given error is exposed in the converted HandlerError, otherwise,
+// a default message with minimal information is attached to the returned error.
+// Roughly taken from https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+// and
+// https://github.com/grpc-ecosystem/grpc-gateway/blob/a7cf811e6ffabeaddcfb4ff65602c12671ff326e/runtime/errors.go#L56.
+func convertServiceError(err error, exposeDetails bool) error {
+	var st *status.Status
+	// Temporal serviceerrors have a Status() method.
+	stGetter, ok := err.(interface{ Status() *status.Status })
+	if !ok {
+		// Not a serviceerror, passthrough.
```

Comment:
What if the error is wrapping a service error?

Reply:
I personally like only checking the outer one. I don't think we should treat errors the same just because they may have the error somewhere in their chain.

Reply:
Good call, switched to `errors.As` instead.

Reply:
Can we just use https://pkg.go.dev/go.temporal.io/api@v1.39.0/serviceerror#ToStatus

Reply:
I was debating what to do here but I think most users will probably wrap just to provide more context but will expect to maintain the semantics.

Reply:
The problem is `serviceerrors ` is thrown low down even the SDK may wrap it at times

Reply:
> I think most users will probably wrap just to provide more context but will expect to maintain the semantics.

I don't necessarily, if they wrap, the expect the behavior of what error they wrap as, not what was wrapped. I shouldn't have to be concerned that somewhere down my chain is a Temporal client error that is now going to affect my Nexus error type.

Reply:
What is the motivation for handling service errors automatically? I don't see it called out in the prerelease docs and it is really only a Go SDK thing AFAIK

Reply:
It's not just a Go SDK thing, it's something all SDKs will need to handle.
If you signal a workflow that's not found from a handler for example, you'll want that operation to fail by default instead of the operation being retried indefinitely.

Reply:
OK, but you don't do that by looking at the `serviceerror`. For example what if a handle tries to get a result from the workflow and the workflow has failed, do you want the handler to be retried indefinitely?

Reply:
shouldn't the Nexus handler be catching `serviceerror` and returning a Nexus error? similar to [this](https://github.com/temporalio/temporal/blob/5bb4d5f5c56694921fe0023859e8ca2c51e5974c/components/nexusoperations/frontend/handler.go#L123)?

but for normal SDK users, who may be writing a `NewSyncOperation` may try to do something like this
```
			err = c.SignalWorkflow(ctx, workflowIDWaitForSignal(input.BusinessID), "", "done", DoneSignal{Done: true})
			if err != nil {
				//TODO: handle &serviceerror.NotFound{Message:"workflow execution already completed"}
				return nil, err
			}
```

but if they don't handle the `serviceerror` explicitly, and just return the error, then it will be treated as a 500 and be retried indefinitely, which probably isn't the behavior a Temporal SDK user would be expecting, and most definitely would not find delightful.

Automatically mapping the `serviceerror` reduces boilerplate in the Temporal SDK user's Nexus handlers, and also doesn't prevent an SDK user from writing their own error handling, it just helps reduce boilerplate and unexpected 500 infinite retries.

Reply:
I am not disagreeing with reducing boilerplate. My issue is the current approach is incomplete and could mislead users. Not all errors returned from the client are `serviceerror` or have `serviceerror` in the error chain. From a users perspective it will look like seemingly random errors are retried and others aren't

Reply:
Hmm... I was looking at random client methods and saw that servicerrors give good coverage.
I can scan the client implementation for other notable errors.

Reply:
Is there a specific error that you have in mind?

I saw `QueryRejectedError` and `WorkflowUpdateServiceTimeoutOrCanceledError`, both of which can be internal errors AFAICT.

Reply:
I think `servicerror` misses some common cases like getting the result of a workflow that has failed would return an error that is not a `servicerror`

If we can clarify the exact desired user experience then we can work backwards to an implementation that gives us that. I think that is really important because other SDKs throw different client errors (we are very inconsistent here across SDKs)

Reply:
> getting the result of a workflow that has failed would return an error

What would this translate to? I think it's still an internal error (the default behavior).

Reply:
If an update is rejected or fails that would also be an error that is not a service error, I would have to double check what error that is.

Reply:
but internal is retryable right? Feels like it should be an `UnsuccessfulOperationError`?

Reply:
> If an update is rejected or fails that would also be an error that is not a service error, I would have to double check what error that is.

Doesn't look like that's returned as an error:

```
		case *updatepb.Outcome_Failure:
			return &ClientPollWorkflowUpdateOutput{
				Error: w.client.failureConverter.FailureToError(v.Failure),
			}, nil
```

Reply:
> but internal is retryable right? Feels like it should be an `UnsuccessfulOperationError`?

Yes, internal is retryable. What do you mean by "it"?

Reply:
Decided to make it non retryable but I couldn't use `UnsuccessfulOperationError` because that's only usable in `Start` and `GetResult`. It's going to be a bad request for lack of better outcome.

Reply:
Also decided to not use `errors.As` and only consider errors directly returned by client methods.
If users wrap the errors, they should decide what error type they want.
Let's get user feedback on this and verify it matches their expectation before we remove the experimental label.

Reply:
Yeah I would really like to get feedback from users here if this is helpful and behaves how they expect.